### PR TITLE
Ministerial reports fixes and improvements

### PR DESCRIPTION
**Pull request by stevehook**
## Context

In order to make ministerial reports production ready we need to fix various discrepancies between the results that TAD generate and our own.

## Changes proposed in this pull request

- [x] Only the original apply 1 and latest apply again applications should be counted in the ministerial applications report.
- [x] Only the original apply 1 and latest apply again applications should be counted for the purpose of evaluating the status of a candidate in the ministerial candidates report. However only the original (apply 1) application is used to evaluate the dominant subject for this report.
- [x] For candidates where a single subject is associated with than 50% of the applications choices that subject is considered the dominant subject for the purposes of the ministerial candidates report. Where that is not the case the candidate is put into the _Split_ subject bucket, e.g. previously if there were 4 choices with 2 having subject A, one subject B and one subject C then subject A was considered the dominant subject. Now that candidate would have not dominant subject and go into the _Split_ bucket.
- [x] Change mappings from `ApplicationChoice#status` to status 'buckets'. (See `MinisterialReport::TAD_STATUS_PRECEDENCE`) to match the way that TAD interpret our statuses.
- [x] Use `ApplicationChoice.current_course_option` rather than `course_option` for ministerial reports (i.e. the course that the candidate was offered rather than the original one they applied to).
- [x] Stop allocating _further education_ level to the _secondary_ subject bucket.
- [x] Add diagnostic code to dump the ids for each status/subject combination in the application and candidate ministerial reports. You need to set the `GENERATE_MINISTERIAL_REPORTS_DIAGNOSTICS` environment variable to
`true` to activate these. They get written to the rails root directory.
- [x] Add rake tasks to take the output from the above and compare them with the results from TAD so that we can more easily work out which candidates in the candidate ministerial report are discrepancies.
- [x] Make the fallback method for picking dominant subject deterministic (when there are multiple subjects for a given course and we cannot pick one based on the course name etc.) We use an explicit sort order rather than depending on a randomly ordered query.
- [x] Stop counting unsuccessful deferrals from last cycle. If an application is deferred from a previous cycle we should include it in this years ministerial reports. However we should not include application choices that were not deferred (e.g. those that ended unsuccessfully in the previous cycle).
- [x] Remove assumption that apply again applications have only a single application choice.
- [x] Various additions to the test data and specs to cover the changes above.

## Guidance to review

- Is the diagnostic code (to dump a list of ids) added to both ministerial reports something we should keep? Does it need to be refined in any way?
- Is the test data comprehensive?

## Link to Trello card

https://trello.com/c/mpp02RdH/4185-epic-make-monthly-report-production-ready

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/services/support_interface/ministerial_report_candidates_export.rb [hunk 1]
```diff
@@ -15,26 +15,43 @@ def call(*)
 
       export_rows[:split] = column_names
 
-      application_forms.find_each do |application|
-        next if application.phase == 'apply_2'
+      subject_report = {}
 
+      application_forms.find_each do |application|
         subjects = determine_subjects(application)
```
Comment (stevehook) [inline app/services/support_interface/ministerial_report_candidates_export.rb:21]: This seems to be always working out the (dominant) subject from the _apply_1_ application (in cases where the candidate has multiple applications). Should it not be doing the same thing as we do below for determining states, i.e. find the latest viable application?

File: app/services/support_interface/ministerial_report_candidates_export.rb [hunk 2]
```diff
@@ -15,26 +15,43 @@ def call(*)
...
 
-        states = if candidate_has_a_successful_apply_2_application?(application)
+        states = if candidate_has_a_viable_apply_2_application?(application)
```
Comment (stevehook) [inline app/services/support_interface/ministerial_report_candidates_export.rb:23]: @duncanjbrown This seems to be the source of some discrepancies. It makes sense to me to pick the latest viable application. This seems to be what is happening in the case of at least one candidate that has 3 applications (and 5 application choices). The latest application choice was `withdrawn` but all the others were `rejected`. TAD seem to be putting this in the `rejected` column whereas we are putting it under `withdrawn`. Isn't that right?

File: app/services/support_interface/ministerial_report_candidates_export.rb (hunk 1 above)
Comment (duncanjbrown) [reply app/services/support_interface/ministerial_report_candidates_export.rb:21]: good q — one for Andrew, I think?

File: app/services/support_interface/ministerial_report_candidates_export.rb (hunk 2 above)
Comment (duncanjbrown) [reply app/services/support_interface/ministerial_report_candidates_export.rb:23]: I think that is right

File: app/lib/ministerial_report.rb [hunk 3]
```diff
@@ -199,7 +197,7 @@ def self.determine_dominant_course_subject_for_report(course)
   def self.determine_dominant_subject_for_report(course_name, course_level, subject_names_and_codes)
     subject_names = subject_names_and_codes.keys
 
-    return :secondary if ['Further education', 'further_education'].include?(course_level)
+    return :further_education if ['Further education', 'further_education'].include?(course_level)
```
Comment (tomas-stefano) [inline app/lib/ministerial_report.rb:200]: Can we move the further education array to a constant?

**Review by tomas-stefano — APPROVED** [review]
🚀 Just a minor comment. But all good.

**Comment by duncanjbrown** [comment]
I've just merged https://github.com/DFE-Digital/apply-for-teacher-training/pull/6470 so we can rebase out a bit from here

File: spec/support/statistics_test_helper.rb [hunk 4]
```diff
@@ -50,6 +50,12 @@ def generate_statistics_test_data
            course_option: course_option_with(level: 'secondary', program_type: 'higher_education_programme', region: 'north_west', subjects: [secondary_subject('Art and design'), secondary_subject('History')]),
            application_form: declined_form)
 
+    withdrawn_form = create(:application_form, :minimum_info, :with_equality_and_diversity_data, sex: 'female', date_of_birth: date_of_birth(years_ago: 80), region_code: :west_midlands, phase: 'apply_1')
```
Comment (duncanjbrown) [inline spec/support/statistics_test_helper.rb:53]: This is fantastic — by definition, I guess we've got all the edge cases now?! Or at least all the ones we and TAD know about. I bet we could make this file a bit more expressive too (later).

**Review by duncanjbrown — APPROVED** [review]
👏 👏 👏  

The hash diffing stuff is awesome, the test data is comprehensive... and the changes are (as far as we know) correct! What more could you want from a PR?

File: app/services/support_interface/ministerial_report_applications_export.rb [hunk 5]
```diff
@@ -10,30 +10,59 @@ def self.run_daily
...
 
       report = choices_with_courses_and_subjects.each_with_object(report) do |choice, report_in_progress|
-        add_choice_to_report(choice, report_in_progress)
+        add_choice_to_report(choice, report_in_progress, subject_ids_report)
       end
 
-      assign_totals_to_report(report)
+      report = assign_totals_to_report(report)
+
+      File.write(
```
[the remark this answers is not in the document]
Comment (stevehook) [reply app/services/support_interface/ministerial_report_applications_export.rb:21]: Add env var as discussed

File: app/services/support_interface/ministerial_report_applications_export.rb [hunk 6]
```diff
@@ -89,6 +94,10 @@ def initialize_empty_report
       report_rows
     end
 
+    def generate_diagnostic_report?
+      ENV['GENERATE_MINISTERIAL_REPORTS_DIAGNOTISTICS'] == 'true'
```
Comment (duncanjbrown) [inline app/services/support_interface/ministerial_report_applications_export.rb:98]: typo! should be GENERATE_MINISTERIAL_REPORTS_DIAGNOSTICS

**Pull request merged**